agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -149,6 +151,186 @@ impl DeriveSchema for Stage {
+                .clone();
+            if let Some(ref depth_field) = graph_lookup.depth_field {
+                let depth_field_schema = Schema::Atomic(Atomic::Long);
+                insert_required_key_into_document(
+                    &mut from_schema,
+                    depth_field_schema,
+                    vec![depth_field.clone()],
+                );
+            }
+            insert_required_key_into_document(
+                &mut state.result_set_schema,
+                Schema::Array(Box::new(from_schema.clone())),
+                vec![graph_lookup.r#as.to_string()],
+            );
+            Ok(state.result_set_schema.to_owned())
+        }
+
+        fn project_derive_schema(
+            project: &ProjectStage,
+            state: &mut ResultSetState,
+        ) -> Result<Schema> {
+            if project
+                .items
+                .iter()
+                .all(|(k, p)| matches!(p, ProjectItem::Exclusion) || k == "_id")
+            {
+                // TODO: handle exclusion $project stages
+                todo!();
+            }
+            // determine if the _id field should be included in the schema. This is the case, if the $project does not have _id: 0.
+            let include_id = project
+                .items
+                .iter()
+                .all(|(k, p)| k != "_id" || !matches!(p, ProjectItem::Exclusion));
+            // project is a map of field name to expression. We can derive the schema for each expression
+            // and then union them together to get the resulting schema.
+            let mut keys = project
+                .items
+                .iter()
+                .filter(|(_k, p)| !matches!(p, ProjectItem::Exclusion))
+                .map(|(k, p)| match p {
+                    ProjectItem::Assignment(e) => {
+                        let field_schema = e.derive_schema(state)?;
+                        Ok((k.to_string(), field_schema))
+                    }
+                    ProjectItem::Inclusion => {
+                        let field_schema = state
+                            .result_set_schema
+                            .get_key(k)
+                            .cloned()
+                            .unwrap_or(Schema::Missing);
+                        Ok((k.to_string(), field_schema))
+                    }
+                    _ => unreachable!(),
+                })
+                .collect::<Result<BTreeMap<String, Schema>>>()?;
+            // if _id has not been excluded and has not been redefined, include it from the original schema
+            if include_id && !keys.contains_key("_id") {
+                keys.insert(
+                    "_id".to_string(),
+                    state
+                        .result_set_schema
+                        .get_key("_id")
+                        .cloned()
+                        .unwrap_or(Schema::Missing),
+                );
+            }
+            Ok(Schema::Document(Document {
```

Comment:
note, if _id is Missing simplify will remove it from required. I change the test code to always call simplify, and when we linked this up to the schema manager we need to make sure we simplify

Reply:
I guess I could just call simplify here